Process gVCF file format natively.

pietropino opened the pull request:
Resolves Issue #2  by implementing in Java the awk workaround for VCF and BED file extraction from gVCF.

deflaux commented:
Thanks for your pull request!  It looks like this may be your first contribution to a Google open source project. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

:memo: **Please visit <https://cla.developers.google.com/> to sign.**

Once you've signed, please reply here (e.g. `I signed it!`) and we'll verify.  Thanks.

---

- If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address.  Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
- If your company signed a CLA, they designated a Point of Contact who decides which employees are authorized to participate. You may need to contact the Point of Contact for your company and ask to be added to the group of authorized contributors. If you don't know who your Point of Contact is, direct the project maintainer to go/cla#troubleshoot.
- In order to pass this check, please resolve this problem and have the pull request author add another comment and the bot will run again.

<!-- need_sender_cla -->

pietropino commented:
I signed the CLA.
Best regards

2018-01-02 23:40 GMT+01:00 Nicole Deflaux <notifications@github.com>:

> Thanks for your pull request! It looks like this may be your first
> contribution to a Google open source project. Before we can look at your
> pull request, you'll need to sign a Contributor License Agreement (CLA).
>
> 📝 *Please visit https://cla.developers.google.com/
> <https://cla.developers.google.com/> to sign.*
>
> Once you've signed, please reply here (e.g. I signed it!) and we'll
> verify. Thanks.
> ------------------------------
>
>    - If you've already signed a CLA, it's possible we don't have your
>    GitHub username or you're using a different email address. Check your
>    existing CLA data <https://cla.developers.google.com/clas> and verify
>    that your email is set on your git commits
>    <https://help.github.com/articles/setting-your-email-in-git/>.
>    - If your company signed a CLA, they designated a Point of Contact who
>    decides which employees are authorized to participate. You may need to
>    contact the Point of Contact for your company and ask to be added to the
>    group of authorized contributors. If you don't know who your Point of
>    Contact is, direct the project maintainer to go/cla#troubleshoot.
>    - In order to pass this check, please resolve this problem and have
>    the pull request author add another comment and the bot will run again.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/verilylifesciences/genomewarp/pull/3#issuecomment-354894540>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AhVOmueG-3ewNns8gTz6CKcLE0WQf6iNks5tGrBOgaJpZM4RRBHq>
> .
>

deflaux commented:
Thanks @pietropino! @cmclean please take a look.

cmclean commented on src/main/java/com/verily/genomewarp/GenomeWarpSerial.java:
```diff
@@ -715,17 +716,32 @@ public static void main(String[] args) {
     List<HomologousRange> namedRegions = null;
     Fasta queryFasta = null, targetFasta = null;
 
+    // Solve Issue #2 - Check if VCS is a gVCF and if so extract 1) variant-only VCF, 2) BED file
```
Users should know whether their data is gVCF or VCF format. I think the processing of the GvcfToVcfAndBed can be simplified if it is only ever invoked on gVCF files, and within GenomeWarpSerial.java there should be argument checking for mutually exclusive flags (i.e., either --raw_query_vcf and --raw_query_bed should both be specified, or a new flag --raw_query_gvcf should be specified, but not both). Then the GvcfToVcfAndBed code is only called if the --raw_query_gvcf flag is passed in.

cmclean commented on src/main/java/com/verily/genomewarp/GenomeWarpSerial.java:
```diff
@@ -715,17 +716,32 @@ public static void main(String[] args) {
     logger.log(Level.INFO, "Creating FASTA structure and jump table");
     try {
-      headerStrings = retrieveVcfHeader(ARGS.rawQueryVcf);
+      headerStrings = retrieveVcfHeader(haveGvcf ? vcfFromGvcf : ARGS.rawQueryVcf);
```
After the flag validation suggested above, I think you can just define a couple new variables to avoid having this logic in multiple places, e.g.

boolean haveGvcf = ARGS.raw_query_gvcf != null;
final String queryVcfToProcess;
final String queryBedToProcess;
if (haveGvcf) {
  queryVcfToProcess = ARGS.workDir + File.separator + "from_gvcf.vcf";
  queryBedToProcess = ARGS.workDir + File.separator + "from_gvcf.bed";
  GvcfToVcfAndBed.saveVcfAndBedFromGvcf(ARGS.rawQueryGvcf, queryVcfToProcess, queryBedToProcess);
} else {
  queryVcfToProcess = ARGS.rawQueryVcf;
  queryBedToProcess = ARGS.rawQueryBed;
}
...

and then just use query{Vcf,Bed}ToProcess in all downstream calls.

cmclean commented on src/main/java/com/verily/genomewarp/utils/GvcfToVcfAndBed.java:
```diff
@@ -0,0 +1,110 @@
+ */
+public class GvcfToVcfAndBed {
+
+  private static final Logger logger = Logger.getLogger(VcfToVariant.class.getName());
```
s/VcfToVariant/GvcfToVcfAndBed/

cmclean commented on src/main/java/com/verily/genomewarp/utils/GvcfToVcfAndBed.java:
```diff
@@ -0,0 +1,110 @@
+   * @param bedFromGvcf intermediate BED path
+   * @return true if query VCF is a gVCF, false otherwise
+   */
+  public static boolean saveVcfAndBedFromGvcf(String rawQueryVcf, String vcfFromGvcf,
```
With other proposed change, please update this name to rawQueryGvcf

cmclean commented on src/main/java/com/verily/genomewarp/utils/GvcfToVcfAndBed.java:
```diff
@@ -0,0 +1,110 @@
+import java.util.logging.Logger;
+
+/**
+ * Helper class for extracting a canonical VCF and a BED file from a gVCF file
```
nit: please end docstring comment with period ".".

cmclean commented on src/main/java/com/verily/genomewarp/utils/GvcfToVcfAndBed.java:
```diff
@@ -0,0 +1,110 @@
+    while ((line = br.readLine()) != null) {
+      // parse line and write it to VCF or BED or none
+      lineNum++;
+      if (line.length() > 0) {
```
This should never be true in a gVCF file, I think this if statement can be omitted.

cmclean commented on src/main/java/com/verily/genomewarp/utils/GvcfToVcfAndBed.java:
```diff
@@ -0,0 +1,110 @@
+          if (fields.length < 8) {
+            throw new IOException("input VCF file has less than 8 columns");
+          }
+          String f5 = fields[4];
```
In general this logic looks great. Clarity would be much improved if some of the logic and magic values were extracted to private functions and member variables. E.g.

private final int MIN_VCF_FIELDS = 8;
private final int START_FIELD_INDEX = 1;
private final int END_FIELD_INDEX = 3;
private final int ALT_BASES_FIELD_INDEX = 4;
private final int FILTER_FIELD_INDEX = 6;
private final int FORMAT_FIELD_INDEX = 7;

private final String UNKNOWN_VALUE = ".";
private final String GVCF_ALT_ALLELE = "<*>";

private boolean isVariantRecord(String altBases) {
  return !UNKNOWN_VALUE.equals(altBases) && !GVCF_ALT_ALLELE.equals(altBases);
}

...

cmclean commented on src/main/java/com/verily/genomewarp/utils/GvcfToVcfAndBed.java:
```diff
@@ -0,0 +1,110 @@
+          if ("PASS".equals(fields[6])) {
+            // BED file - compute line
+            try {
+              Long start = Long.parseLong(fields[1]) - 1L;
```
Please add a comment noting that VCF is 1-based full closed indexing, whereas BED is 0-based half open indexing, hence the -1L here.

cmclean replied:
Please abstract this into a separate function for clarity:

private String bedLineFromVcfFields(String[] fields)

cmclean commented on src/main/java/com/verily/genomewarp/utils/GvcfToVcfAndBed.java:
```diff
@@ -0,0 +1,110 @@
+              Long start = Long.parseLong(fields[1]) - 1L;
+              Long end = start + fields[3].length();
+              String f8 = fields[7];
+              if (!".".equals(f8)) {
```
This if statement is unnecessary, the split on ";" would just return a 1-element array containing "." in that case.

cmclean commented on src/main/java/com/verily/genomewarp/utils/GvcfToVcfAndBed.java:
```diff
@@ -0,0 +1,110 @@
+   */
+  public static boolean saveVcfAndBedFromGvcf(String rawQueryVcf, String vcfFromGvcf,
+      String bedFromGvcf) throws IOException {
+    PrintWriter outBed = new PrintWriter(Files.newBufferedWriter(Paths.get(bedFromGvcf), UTF_8));
```
Please use a try-with-resources (https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) that includes all three of these resources to ensure all files get closed properly.

cmclean commented on src/main/java/com/verily/genomewarp/utils/GvcfToVcfAndBed.java:
```diff
@@ -0,0 +1,110 @@
+   * @return true if query VCF is a gVCF, false otherwise
+   */
+  public static boolean saveVcfAndBedFromGvcf(String rawQueryVcf, String vcfFromGvcf,
+      String bedFromGvcf) throws IOException {
```
Given the updated contract where this expects a valid gVCF as input, upon non-successful execution (either failure to parse or IOException reading/writing) please delete the temp files and then just return false. Then GenomeWarpSerial can throw an exception indicating unsuccessful execution.

pietropino commented:
Hello @cmclean, thanks for your prompt feedback.
Your amendments have been committed on the same branch. Besides, I've improved the user-provided paths validation, refactored the first two tests, and added a new one for the malformed gVCF case.

cmclean commented on src/main/java/com/verily/genomewarp/GenomeWarpSerial.java:
```diff
@@ -717,31 +720,43 @@ public static void main(String[] args) {
     if (haveGvcf) {
+      queryVcfToProcess = ARGS.workDir + File.separator + "from_gvcf.vcf";
+      queryBedToProcess = ARGS.workDir + File.separator + "from_gvcf.bed";
+      if (ARGS.rawQueryVcf != null || ARGS.rawQueryBed != null) {
```
Nit: I think the flow is a little clearer if you pull the flag validation performed here and in lines 741-744 above all this processing rather than interleaving it with the computation. E.g.

// Query input flag validation.
if (ARGS.rawQueryGvcf != null && (ARGS.rawQueryVcf != null || ARGS.rawQueryBed != null)) {
  fail("Arguments (--raw_query_vcf, --raw_query_bed) and --raw_query_gvcf are mutually exclusive");
}
if ((ARGS.rawQueryVcf == null) != (ARGS.rawQueryBed == null)) {
  fail("Either both or neither of --raw_query_vcf and --raw_query_bed must be specified");
}
if (ARGS.rawQueryGvcf == null && ARGS.rawQueryVcf == null) {
  fail("Either (--raw_query_vcf, --raw_query_bed) or --raw_query_gvcf must be specified");
}

Then you can safely operate with the haveGvcf boolean defined as is.

cmclean commented on src/main/java/com/verily/genomewarp/utils/GvcfToVcfAndBed.java:
```diff
@@ -19,92 +19,145 @@
+  private static final String PASS_FILTER_VALUE = "PASS";
+  private static final String FORMAT_FIELD_SEP_REGEX = ";";
+  private static final String FORMAT_END_KEY = "END";
+  private static final int CHROM_FIELD_INDEX = 0;
```
nit: put this between current lines 34-35 so the field indices all appear together and in order.

cmclean commented on src/main/java/com/verily/genomewarp/utils/GvcfToVcfAndBed.java:
```diff
@@ -19,92 +19,145 @@
+            // not a header - choose destination according to columns
+            String[] fields = line.split(FIELD_SEP_REGEX);
+            if (fields.length < MIN_VCF_FIELDS) {
+              throw new IOException("input VCF file has less than " + MIN_VCF_FIELDS + " columns");
```
Consider adding lineNum to this error too?

cmclean commented on src/main/java/com/verily/genomewarp/utils/GvcfToVcfAndBed.java:
```diff
@@ -19,92 +19,145 @@
+            if (fields.length < MIN_VCF_FIELDS) {
+              throw new IOException("input VCF file has less than " + MIN_VCF_FIELDS + " columns");
+            }
+            String f5 = fields[ALT_BASES_FIELD_INDEX];
```
Please use a more descriptive name, e.g.

String altBases = ...

cmclean commented on src/main/java/com/verily/genomewarp/utils/GvcfToVcfAndBed.java:
```diff
@@ -19,92 +19,145 @@
-          String f5 = fields[4];
-          if (!".".equals(f5) && !"<*>".equals(f5)) {
-            // variant VCF file
+    try {
```
I think the nesting of trys can be avoided and increase clarity.

Path bedFromGvcfPath;
...
try {
  bedFromGvcfPath = Paths.get(bedFromGvcf);
  ...
} catch (Exception ex) {
  logger.log(Level.SEVERE, "Error creating paths: " + ex.toString());
  return false;
}

try (PrintWriter ...) {
  ...
}

cmclean replied:
(Or just put the Paths.get calls within the try-with-resources block so they get caught there)

pietropino replied:
I'll use the first option (like I did originally, don't know why I changed =) ), since I need the intermediate files paths available outside the try scope when I remove them on error.

cmclean commented on src/main/java/com/verily/genomewarp/utils/GvcfToVcfAndBed.java:
```diff
@@ -19,92 +19,145 @@
+      String bedFromGvcf) {
     String line;
-    boolean isGvcf = false;
+    boolean isSuccessfulGvcf = false;
```
How about

boolean successfullyParsedGvcf = true;

And then only set the boolean to false within catch blocks?

pietropino replied:
I can refactor the name but I need it to be true only when I have at least one BED line.

cmclean replied:
Why? If a gVCF only contains non-passing variants the expected result is a VCF with the variant calls and an empty BED file. Looking through the downstream code I believe it handles an existing but empty BED file appropriately (nothing gets transformed).

pietropino replied:
OK then, I was in doubt whether an empty BED file was a valid use case.

cmclean commented on src/main/java/com/verily/genomewarp/utils/GvcfToVcfAndBed.java:
```diff
@@ -19,92 +19,145 @@
+    // hence the -1L here.
+    Long start = Long.parseLong(fields[START_FIELD_INDEX]) - 1L;
+    Long end = start + fields[END_FIELD_INDEX].length();
+    String f8 = fields[FORMAT_FIELD_INDEX];
```
Please use more descriptive names here, e.g. String rawFormatField, String[] formatFieldTokens

cmclean commented on src/test/java/com/verily/genomewarp/utils/GvcfToVcfAndBedTest.java:
```diff
@@ -34,65 +35,81 @@
+        BufferedReader gotVcfBR = Files.newBufferedReader(Paths.get(outVcf), UTF_8)) {
+      while ((line1 = gotVcfBR.readLine()) != null) {
+        line2 = expVcfBR.readLine();
+        assertTrue(line2 != null);
```
Is this assertTrue needed given the assertEquals on the next line?

cmclean replied:
If so, please static import and use assertNotNull

cmclean commented on src/test/java/com/verily/genomewarp/utils/GvcfToVcfAndBedTest.java:
```diff
@@ -34,65 +35,81 @@
+        assertTrue(line2 != null);
+        assertEquals(line1, line2);
+      }
+      assertTrue(expVcfBR.readLine() == null);
```
assertNull(expVcfBR.readLine());

cmclean commented on src/test/java/com/verily/genomewarp/utils/GvcfToVcfAndBedTest.java:
```diff
@@ -34,65 +35,81 @@
+        BufferedReader expVcfBR =
+            Files.newBufferedReader(Paths.get(expVcf.getCanonicalPath()), UTF_8);
+        BufferedReader gotVcfBR = Files.newBufferedReader(Paths.get(outVcf), UTF_8)) {
+      while ((line1 = gotVcfBR.readLine()) != null) {
```
Optional: You could extract the logic of lines 63-68 into a private helper function assertBufferedReadersEqual(BufferedReader r1, BufferedReader r2) since the identical logic is used for the VCF and BED files.

pietropino commented:
@cmclean thanks for your stimulating and instructive remarks! The revised code is now available on the same branch. 
Cheers.

pietropino commented:
Thanks to you! It's been useful for me too, I've learned new things, also about genetics.
